File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
node_cnt = 1

def dfs(graph, v, visited):
    global node_cnt
    
    for nv in graph[v]:
        if not visited[nv]:
            node_cnt += 1
            visited[nv] = True
            dfs(graph, nv, visited)

def solution(n, wires):
    global node_cnt
    answer = 1e9
    
    for i in range(n - 1):
        graph = [[] for _ in range(n + 1)]
        visited = [False for _ in range(n + 1)]
        nodes = []
        
        for w in range(len(wires)):
            if w == i:
                continue
            
            v1, v2 = wires[w]
            graph[v1].append(v2)
            graph[v2].append(v1)
        
        for j in range(1, n + 1):
            if not visited[j]:
                visited[j] = True
                node_cnt = 1
                dfs(graph, j, visited)
                nodes.append(node_cnt)
                
        diff = abs(nodes[0] - nodes[1])
        answer = min(answer, diff)
            
    return answer

File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    static ArrayList<ArrayList<Integer>> graph;
    static boolean[] visited;
    static int nodeCnt = 1;
    static int answer = Integer.MAX_VALUE;
    
    public int solution(int n, int[][] wires) {
        for(int i = 0; i < wires.length; i++){
            graph = new ArrayList<>();
            visited = new boolean[n + 1];
            int[] towerCnt = new int[2];
            int idx = 0;
            
            for(int j = 0; j <= n; j++){
                graph.add(new ArrayList<>());
            }
            
            for(int j = 0; j < wires.length; j++){
                if(i == j) continue;
                
                graph.get(wires[j][0]).add(wires[j][1]);
                graph.get(wires[j][1]).add(wires[j][0]);
            }
            
            for(int j = 1; j <= n; j++){
                if(!visited[j]){
                    visited[j] = true;
                    nodeCnt = 1;
                    dfs(j);
                    
                    towerCnt[idx++] = nodeCnt;
                }
            }
            
            answer = Math.min(answer, Math.abs(towerCnt[0] - towerCnt[1]));
        }
        
        return answer;
    }
    
    public static void dfs(int startNode){
        for(int i = 0; i < graph.get(startNode).size(); i++){
            int adjNode = graph.get(startNode).get(i);
            
            if(!visited[adjNode]){
                nodeCnt++;
                visited[adjNode] = true;
                dfs(adjNode);
            }
        }
    }
}